charmhelpers/contrib/openstack/context.py

```diff
@@ -2988,3 +2988,91 @@ def get_ovs_portdata(self, bond):
                 'lacp-time': bond_config['lacp-time'],
             },
         }
+
+
+class SRIOVContext(OSContextGenerator):
+    """Provide context for configuring SR-IOV devices."""
+
+    def __init__(self, numvfs_key=None, device_mappings_key=None):
+        """Initialize map from PCI devices and configuration options.
+
+        :param numvfs_key: Config key for numvfs (default: 'sriov-numvfs')
+        :type numvfs_key: Optional[str]
+        :param device_mappings_key: Config key for device mappings
+                                    (default: 'sriov-device-mappings')
+        :type device_mappings_key: Optional[str]
+        """
+        numvfs_key = numvfs_key or 'sriov-numvfs'
+        device_mappings_key = device_mappings_key or 'sriov-device-mappings'
+
+        devices = pci.PCINetDevices()
+        charm_config = config()
+        sriov_numvfs = charm_config.get(numvfs_key) or ''
+        sriov_device_mappings = charm_config.get(device_mappings_key) or ''
+
+        # create list of devices from sriov_device_mappings config option
+        sriov_mapped_devices = [
+            pair.split(':', 1)[1]
+            for pair in sriov_device_mappings.split()
+        ]
+
+        # create list of devices from sriov_numvfs config option
+        sriov_config_devices = {}
+        try:
+            for ifname, numvfs in (
+                    pair.split(':', 1)
+                    for pair in sriov_numvfs.split()):
+                sriov_config_devices.update({ifname: numvfs})
```

Comment:
Perhaps we could re-write this as:

```python
sriov_config_devices = {
    ifname: numvfs for ifname, numvs in (
        pair.split(':', 1) for pair in sriov_numvfs.split())}
```

Reply:
Ah, cool, I did try to wrangle it into dict comprehension but failed, so thank you for pointing me in the direction of how.